agent/tcs/client/client.go

```diff
@@ -149,30 +147,85 @@ func (cs *clientServer) publishMetrics() {
 
 // publishMetricsOnce is invoked by the ticker to periodically publish metrics to backend.
 func (cs *clientServer) publishMetricsOnce() {
-	metadata, taskMetrics, err := cs.statsEngine.GetInstanceMetrics()
+	// Get the list of objects to send to backend.
+	requests, err := cs.metricsToPublishMetricRequests()
 	if err != nil {
 		log.Warn("Error getting instance metrics", "err", err)
-		return
 	}
 
-	if *metadata.Idle {
-		// Idle instance, send message and return.
-		cs.MakeRequest(ecstcs.NewPublishMetricsRequest(metadata, taskMetrics))
-		return
+	// Make the publish metrics request to the backend.
+	for _, request := range requests {
+		cs.MakeRequest(request)
 	}
+}
 
+// metricsToPublishMetricRequests gets task metrics and converts them to a list of PublishMetricRequest
+// objects.
+func (cs *clientServer) metricsToPublishMetricRequests() ([]*ecstcs.PublishMetricsRequest, error) {
+	metadata, taskMetrics, err := cs.statsEngine.GetInstanceMetrics()
+	if err != nil {
+		return nil, err
+	}
+
+	var requests []*ecstcs.PublishMetricsRequest
+	if *metadata.Idle {
+		fin := true
+		metadata.Fin = &fin
+		// Idle instance, we have only one request to send to backend.
+		requests = append(requests, ecstcs.NewPublishMetricsRequest(metadata, taskMetrics))
+		return requests, nil
+	}
 	var messageTaskMetrics []*ecstcs.TaskMetric
-	for i := range taskMetrics {
-		messageTaskMetrics = append(messageTaskMetrics, taskMetrics[i])
+	numTasks := len(taskMetrics)
+
+	for i, taskMetric := range taskMetrics {
+		messageTaskMetrics = append(messageTaskMetrics, taskMetric)
+		var fin bool
+		if (i + 1) == numTasks {
+			// If this is the last task to send, set fin to true
+			fin = true
+		} else {
+			fin = false
+		}
+		requestMetadata := fromMetricsMetadata(metadata, fin)
 		if (i+1)%tasksInMessage == 0 {
 			// Construct payload with tasksInMessage number of task metrics and send to backend.
-			cs.MakeRequest(ecstcs.NewPublishMetricsRequest(metadata, messageTaskMetrics))
+			requests = append(requests, ecstcs.NewPublishMetricsRequest(requestMetadata, copyTaskMetrics(messageTaskMetrics)))
 			messageTaskMetrics = messageTaskMetrics[:0]
 		}
 	}
 
 	if len(messageTaskMetrics) > 0 {
-		// Send remaining task metrics to backend.
-		cs.MakeRequest(ecstcs.NewPublishMetricsRequest(metadata, messageTaskMetrics))
+		fin := true
+		requestMetadata := fromMetricsMetadata(metadata, fin)
+		// Create a request with remaining task metrics.
+		requests = append(requests, ecstcs.NewPublishMetricsRequest(requestMetadata, messageTaskMetrics))
 	}
+	return requests, nil
+}
+
+// fromMetricsMetadata creates a new MetricsMetadata object from a given MetricsMetadata objecy.
```

Comment:
Nit, 'objecy' typo.

I'm also not a fan of this function. It seems to me like it would better be a `deepCopyMetricsMetadata` function with the caller setting `fin` themselves. Even if this continues to set `fin`, I think `copy` more clearly conveys what it does.
It's possible I'm misunderstanding the intent of this though.